source/extensions/filters/network/rocketmq_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,244 @@
+  Tcp::ConnectionPool::Cancellable* cancellable = connectionPool->newConnection(*upstream_request_);
+  if (cancellable) {
+    handle_ = cancellable;
+    ENVOY_LOG(trace, "No connection is available for now. Create a cancellable handle. Opaque: {}",
+              opaque);
+  } else {
+    /*
+     * UpstreamRequest#onPoolReady or #onPoolFailure should have been invoked.
+     */
+    ENVOY_LOG(trace,
+              "One connection is picked up from connection pool, callback should have been "
+              "executed. Opaque: {}",
+              opaque);
+  }
+}
+
+RouterImpl::UpstreamRequest::UpstreamRequest(RouterImpl& router,
+                                             Tcp::ConnectionPool::Instance& connectionPool)
+    : router_(router), connectionPool_(connectionPool) {
+  (void)connectionPool_;
+}
+
+void RouterImpl::UpstreamRequest::onPoolReady(Tcp::ConnectionPool::ConnectionDataPtr&& conn,
+                                              Upstream::HostDescriptionConstSharedPtr host) {
+  router_.connection_data_ = std::move(conn);
+  router_.upstream_host_ = host;
+  router_.connection_data_->addUpstreamCallbacks(router_);
+  if (router_.handle_) {
+    ENVOY_LOG(trace, "#onPoolReady, reset cancellable handle to nullptr");
+    router_.handle_ = nullptr;
+  }
+  // TODO: we may optimize out encoding in case we there is no protocol translation.
+  Buffer::OwnedImpl buffer;
+  Encoder::encode(router_.active_message_->downstreamRequest(), buffer);
+  router_.connection_data_->connection().write(buffer, false);
+  ENVOY_LOG(trace, "Write data to upstream OK. Opaque: {}",
+            router_.active_message_->downstreamRequest()->opaque());
+
+  if (router_.active_message_->isOneway()) {
+    ENVOY_LOG(trace,
+              "Reset ActiveMessage since data is written and the downstream request is one-way. "
+              "Opaque: {}",
+              router_.active_message_->downstreamRequest()->opaque());
+    router_.reset();
+  }
+}
+
+void RouterImpl::UpstreamRequest::onPoolFailure(Tcp::ConnectionPool::PoolFailureReason reason,
+                                                Upstream::HostDescriptionConstSharedPtr host) {
+  switch (reason) {
+  case Tcp::ConnectionPool::PoolFailureReason::Overflow: {
+    ENVOY_LOG(error, "Unable to acquire a connection to send request to upstream");
+    router_.active_message_->onError("overflow");
+  } break;
+
+  case Tcp::ConnectionPool::PoolFailureReason::RemoteConnectionFailure: {
+    ENVOY_LOG(error, "Failed to make request to upstream host {} due to connection pool error",
+              host->address()->asString());
+    router_.active_message_->onError("remote connection failure");
+  } break;
+
+  case Tcp::ConnectionPool::PoolFailureReason::LocalConnectionFailure: {
+    ENVOY_LOG(error, "Failed to make request to upstream due to connection pool error. Host: {}",
+              host->address()->asString());
+    return;
```

Comment:
If this error occurs, what is our follow-up process? Why not make a local reply here.

Reply:
fixed already.